src/v/cluster/rm_stm.cc

```diff
@@ -1502,13 +1502,17 @@ ss::future<bool> rm_stm::sync(model::timeout_clock::duration timeout) {
     auto ready = co_await persisted_stm::sync(timeout);
     if (ready) {
         if (_mem_state.term != _insync_term) {
-            // There is a term change, issue a checkpoint purge to followers
-            // All other replicate ops need to wait until this finishes, this is
-            // so that the followers clean up the state before we start applying
-            // changes to the in memory state.
-            co_await _checkpoint_applied_replicate.with(
-              [this]() { return replicate_checkpoint_applied(_insync_term); });
-            _mem_state = mem_state{.term = _insync_term};
+            // There is a term change and we are the new leader. Two main steps
+            // here.
+            // 1. Reconcile the mem state: Promotes the parked checkpoint state
+            // into _mem_state.
+            // 2. Notify all the new followers that the checkpoint has been
+            // applied before we
+            //    allow clients to progress.
+            co_await _checkpoint_applied_replicate.with([this]() {
+                reconcile_mem_state();
```

Comment:
yeah, it's better to put the check here to avoid to checking it both in `reconcile_mem_state` and `replicate_checkpoint_applied`:

```
if (_parked_checkpointed_mem_state) {
    reconcile_mem_state();
    co_await replicate_checkpoint_applied()
}
```

It reduces the combination of branches from 4 to 2 so it's easier to understand what's happening.